ENH: Generalized ufunc signature expansion for frozen and flexible dimensions

mhvk opened the pull request:
EDIT: now the replacement of #11132

An alternative to #11132, much more following its logic, but (I think) clearer. Mostly for @mattip to look at. I like the use of flags, although it may seem a bit overkill if we have just one. But I do hope to have a broadcast option (tentatively `n|1`). For now, I also kept the frozen dimensions, as those are quite easy.

jaimefrio committed:
ENH: Add frozen dimensions to gufunc signatures

mattip committed:
ENH: convert reserved field to version, set new version to 1

So that we can keep track of expansions of the PyUFuncObject.

mattip committed:
ENH: test, document, implement flexible signature.

Goal is allow signatures like (m?,n),(n,p?)->(m?,p?) for matmul.

mhvk committed:
MAINT: Use flags instead of flexible_dims argument.

This allows future expansion and is arguably easier to follow.

Also adds tests, of the signature parsing, of the ignoring of
dimensions, and of use of the frozen dimensions via cross1d.

mhvk committed:
MAINT: Move dimension logic to its own routine; simplify.

It got very long and now matches logically the later call to get
core sizes.

mhvk committed:
MAINT: Remove n_dims_used, and reuse core_num_dims instead.

mhvk committed:
MAINT: Ensure signatures with frozen sizes written differently compare equal.

Was comparison by the string representation so that "03" and "3" would
be considered different dimensions.  That would be surprising for implementing
an inner loop...

charris commented:
Notifying @shoyer and @mrocklin for these discussions, as they are looking to expand on the `array_ufunc` idea and this is in the same general area.

mhvk commented:
To get a full sense of what the ideas are, have a look at #11179 (just to be sure, all based on @mattip's work). The test cases may right now provide the best documentation...

charris commented:
Should have @njsmith here also.

mattip commented:
After reading the [comment](https://github.com/numpy/numpy/pull/11132#issuecomment-392540951) in #11132 it seems you want to first implement fixed core dimensions and then add flexible ones?

mhvk commented:
Ah, partly an answer to my question in #11132 - yes, starting with `?` and fixed sizes is fine too - #11132 was just to show how easy it was to add broadcasting.

mattip commented:
@mhvk could you rebase this off master, or do you want me to do it? Also notice there are changes to umath/scalar.c.src which add `matrix_multiply`. They belong in the final matmul pr #11133 and should be removed from this PR

mhvk commented:
I'm in the process of rebasing - I'll remove the matmul specific commit.

mhvk commented:
OK, I rebased this and think it is now ready for review. In rebasing, I used @jaimefrio's original commit for frozen dimensions, so that it is still attributed properly. I similarly used @mattip's commits, but squashed to 2. Note that I rebased on top of #11173 and #11176, to avoid having another difficult rebase after those have been merged.

One note: the doc changes are just @mattip's - I've made more elaborate changes in the broadcasting follow-up, but hoped not to have to split the relevant parts out.

mhvk committed:
MAINT: Ensure correct order of inner strides for ignored dimensions.

Previously, when a core dimension at the end was missing, the strides
were not re-ordered accordingly, and the inner loop code needed to
check whether the corresponding dimension was zero. But if the strides
are passed on correctly, this is not needed at all, and the dimension
can just be unity (not indicating broadcasting this time).

mhvk commented:
Hah, the failing test from 32 bit exposed an interesting bug: previously, @mattip had implemented the flexible dimensions such that the elementary function had to check whether a dimension was zero, and then swap strides. But with matrix-multiply, this changed the behaviour of one of the test cases, in which it is called with `matrix_multiply(np.ones((0, 10)), np.ones((10, 0)))`, i.e., with real zeros for the dimensions.

So, this was clearly fragile. But fortunately also not needed: by just passing on the strides in the correct order, this is solved and, as one would like, the elementary function doesn't have to know about whether flexible dimensions are being used. Much nicer!

A nice side benefit is that `matmul` now becomes even more trivial to implement, as its code does not have to change at all: it just needs wrapping as a `gufunc`.

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2199,10 +2362,10 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     int broadcast_ndim, iter_ndim;
-    int core_num_dims_array[NPY_MAXARGS];
-    int *core_num_dims;
+    int core_num_dims[NPY_MAXARGS];
```
As per discussion in PR #11176, this should be renamed

mhvk replied:
Need to think a bit about the logic of the names, since we're copying and adjusting two arrays: `core_num_dims` and `core_dim_sizes`. In both cases, the copies represent the *actual* number of dimensions that the operands have and the sizes that they imply. So, long versions could be
`actual_core_num_dims` and `actual_core_dim_sizes` - or remove `_core`.

Alternatively, I have been wondering whether it would make sense to have a mini-struct `actual` that had whatever parts of the ufunc would need changing (could be expanded to include `core_dim_ixs` for "calculated" output indices, e.g.). So that would mean `actual->core_num_dims`, `actual->core_dim_sizes`. But this could be done in a separate PR as well.

mhvk replied:
p.s. I'm also not so happy that as is, `core_num_dims` gets copied even if in standard usage it never gets adjusted. But I guess it is only a few numbers, so very little overhead...

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2538,7 +2715,7 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     core_dim_ixs_size = 0;
     for (i = 0; i < nop; ++i) {
-        core_dim_ixs_size += core_num_dims[i];
+        core_dim_ixs_size += ufunc->core_num_dims[i];
```
Should this be precalclulated and stored as part of the ufunc struct?

mhvk replied:
I wondered about that, but it seems a bit excessive. Another thing one might do is to store it at the end of `core_offsets`, so that that really equal `cumsum(core_num_dims)`

mhvk replied:
p.s. If we want to go this route, I do think this should be a different PR -- as long as we don't have a release, we are free to change the struct even without a version number change. And my overall sense is that it is rarely really needed: even now, one can just do
```
core_dim_ixs_size = ufunc->core_offsets[ufunc->nargs - 1] + ufunc->core_num_dims[ufunc->nargs - 1]
```

mattip commented:
tests for fixed dim signature still missing? The `cross1d` function is never used in tests AFAICT.

mhvk commented:
Rebased to get rid of conflicts, and now including tests of `cross1d`.

mattip commented:
@charris, @njsmith,  Do you have a sense of how controversial allowing flexible/fixed gufunc core signatures is? Can we go ahead and merge this new feature soonish or does it need another round on the mailing list? Note this is the enabler for changing `matmul` PR #11133, which needs more work

If it is good to go, it should get an entry in `Improvements` in `doc/release/1.15.0-notes.rst`

mattip commented on doc/source/reference/c-api.generalized-ufuncs.rst:
```diff
@@ -139,22 +140,35 @@ Notes:
 #. White spaces are ignored.
+#. An integer as a dimension name freezes that dimension to the value,
+#. The name can be suffixed with a question mark, this make the dimension a
+   core dimension only if it exists on the input or output, otherwise 1 is used.
```
We should make clear that the 1 will not be magically inserted if the core dimensions do not match

"otherwise 1 is used for a missing core dimension" or so

mattip commented on doc/source/reference/c-api.generalized-ufuncs.rst:
```diff
@@ -139,22 +140,35 @@ Notes:
++-------------+----------------------------+-----------------------------------+
+| dot2d       | ``(m?,n),(n,p?)->(m?,p?)`` | all four of the above at once     |
++-------------+----------------------------+-----------------------------------+
+| cross1d     | ``(3),(3)->(3)``           | cross product where last dim is 3 |
```
"is" => "must be"

mhvk commented:
Sorry, that should be https://gist.github.com/mhvk/a3c17f0db77b91151d50a0d278621d97, as I should at least include one function that needs the fix for when there are only output arguments with a core dimension of 3 (`fk5hip`).

mattip commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -209,10 +209,35 @@ typedef struct _tagPyUFuncObject {
+         *       UFUNC_CORE_FROZEN for a numerical dimension
+         */
+        npy_uint32 *core_dim_flags;
+
```
The changes need an update to doc/source/reference/c-api.types-and_structures.rst, line 680 or so, with a qualifier that the new fields should only be used if `version > 0`. Come to think of it, we should be checking that as well, just in case someone is mixing headers from old code

mhvk replied:
OK, I see now. From your comments below, you're specifically worried about C code that was compiled with an old include file, and then will try to construct the ufunc with parse_signature but with an old version. Seems like that should be do-able.

mhvk replied:
But it would need a test - would it be enough to just use a new `ufunc` but with version = 0 and then checking the new additions do not get set?

mattip replied:
Sounds good to me. If you want I can work on this

mhvk replied:
Yes, super! I've spent a bit more time on this than I really should have, and things related to my normal work have been piling up...

mhvk replied:
(Commenting here instead of at https://github.com/mhvk/numpy/pull/1 so that history get kept.)

I'm still a bit confused when exactly this becomes a problem. Suppose I wrote my own ufunc module and have it compiled with the old `ufuncobject.h`, so that internally it counts on the struct not to have `core_dim_flags` and `core_dim_sizes`. Now if I load my module with a new numpy, what exactly happens? From what I saw in cases where I mistakenly loaded astropy in my development branch, it seems one does get a warning automatically, but then code gets run.

Now, presumably, this old module is organized a bit like `_umath_tests.src.c` and thus has a `PyMODINIT_FUNC`, which will run several `PyUFunc_FromFuncAndDataAndSignature`. These in turn will call `parse_signature` and, hey, presto, segmentation fault if that writes to outside the old ufunc struct. So, it does seem one needs a version check in `parse_signature` in addition to having one here that guards against accessing the new additions.

mattip replied:
hmm. `_parse_signature` is marked `static` so it should never leak out of `PyUFunc_FromFuncAndDataAndSignature`, and the latter allocates its own `PyUFuncObject`. So AFAICT we cannot enter `_parse_signature` with an old struct. My bad, I should have thought this through before requesting changes to this code.

mhvk replied:
Well, that at least is good news!

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -437,26 +469,70 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
+                frozen_size = UFUNC_CORE_DIM_SIZE_UNSET;
+            }
+            else {
+                frozen_size = _get_size(signature + i);
```
only if `ufunc->version > 0`

mattip replied:
version here can never be 0, ignore this comment

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -437,26 +469,70 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
+             * get its index, and set its properties
+             */
+            for(ix = 0; ix < ufunc->core_num_dim_ix; ix++) {
+                if (frozen_size > 0 ?
```
what if ufunc->version == 0?

mattip replied:
version here can never be 0, ignore this comment

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -437,26 +469,70 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
+                }
+                if (can_ignore) {
+                    ufunc->core_dim_flags[ix] |= UFUNC_CORE_DIM_CAN_IGNORE;
+                }
```
what happens if ufunc->version == 0?

mattip replied:
version here can never be 0, ignore this comment

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -493,7 +569,12 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
     ufunc->core_dim_ixs = PyArray_realloc(ufunc->core_dim_ixs,
-            sizeof(int)*cur_core_dim);
+            sizeof(int) * cur_core_dim);
+    ufunc->core_dim_sizes = PyArray_realloc(ufunc->core_dim_sizes,
```
ignored if ufunc->version == 0

mattip replied:
version here can never be 0, ignore this comment

mattip commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -406,13 +430,21 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
+
+    if (ufunc->core_num_dims == NULL || ufunc->core_dim_ixs == NULL ||
+        ufunc->core_offsets == NULL || ufunc->core_dim_flags == NULL ||
+        ufunc->core_dim_sizes == NULL) {
```
what happens if ufunc->version == 0?

mattip replied:
version here can never be 0, ignore this comment

mattip committed:
DOC: updates and fixes for gufunc descriptions.

mhvk committed:
DOC: Release notes for frozen dimensions.

mhvk committed:
DOC: Release note for flexible dimensions.

mattip commented on doc/release/1.15.0-notes.rst:
```diff
@@ -416,5 +416,29 @@ differently from variable ones indicated with a letter; the loop still is
+are very similar to generalized ufuncs in that they operate over core
+dimensions, but one could not present them as such because they were able to
+deal with inputs in which a dimension is missing. To support this, it is now
+allowed to postfix a dimension name with a question mark to indicate that that
```
nit: that that -> that the

mhvk replied:
done

mattip commented:
@mhvk should I go ahead and create the `_initialize_core_information(ufunc, core_num_dims, core_dim_sizes, core_dim_flags)` function or will you do it?

mhvk commented:
@mattip - I think the new function is useful, but with the worry about `version` out of the way, it probably makes more sense to do it as a follow-up to this PR. Indeed, that would also allow us to think a bit more about other uses and implementation. So, I opened a new issue for it: #11228.

I will change my flags definitions here to make this easier...

mhvk commented:
OK, changed the flags to ensure that they all indicate something is not fixed, whether it is the size or the number of dimensions. Also fixed up a few small nits in the docs.

mhvk committed:
MAINT: factor out initialization of variable parts of the ufunc.

That is, parts like core_num_dims which can get changed during
execution, e.g., because keepdims is set or a dimension is flexible.

mhvk commented:
@mattip - yes, might as well. I combined it with an initialization function after all; this now also checks that the version number is 0 or 1.

mhvk committed:
MAINT: change the name of the actual core_num_dims.

Now op_core_num_dims to make clear it is the number of core dimensions
an operand has, rather than that assumed by the ufunc.

mhvk commented:
@mattip - I now also changed the name of `core_num_dims` to `op_core_num_dims` to (hopefully) suggest more clearly that it is the number of core dimensions of the operand, rather than what is assumed by the ufunc. I felt that for `core_dim_sizes` and `core_dim_flags` no change in name was necessary.

mattip reviewed (approved):
LGTM. Have we reached a consensus on the mailing list that this is a change we are comfortable with?

mattip commented:
Next up - rebasing and completing the matmul-as-ufunc PR #11133

charris commented:
Needs rebase. Is it the consensus that this is the way to go?

charris commented:
@mhvk Thanks for the warning. If this touches the ufunc structure I'd rather push it off, there is already a lot of churn in 1.15, more than I'd like. That said, it would also be good to get this stuff in a stable state before we drop Python 2.7.

mattip commented:
cherry-picked 4e2f72a from pr #11133

mhvk commented:
@mattip - not sure it is a good idea to do the change in name here - certainly I have code that actually sets `->legacy_inner_loop_selector` - should we really be forcing people to change names? It seems a bad idea to me; much more disruptive than adding new fields which, if untouched, do just the right thing.

In any case, I think this should be separate from this PR!

mhvk commented:
@mattip - I force-pushed back the old version, to remove the commit and its reversal.

mattip commented:
Thanks. Sorry for the noise.

mhvk commented:
@mattip - I think there was consensus that this one could go in. Shall we do it so you can proceed with `matmul`?

shoyer commented:
@mhvk I was just about to say the same! But maybe we'll wait until after the NumPy Birds-of-a-Feather meeting at SciPy in an hour :).

In case you're curious, here are my slides giving an overview of currently active NEPs: https://docs.google.com/presentation/d/1NE9ahlCC5dvhUJOWcpG84VxiGzoenGMdbrQLV10o_n4/edit?usp=sharing

mattip commented:
Since PR #11429 accepting NEP 20 was merged, this is ready to go.

mattip reviewed (changes requested):
Release notes need to go to 1.16 not 1.15

mhvk commented:
Oops! Now corrected (and rebased in the process)

mattip commented:
It would be good to get another review of this before merging, I don't think we have a clearly stated requirement for two reviewers but it seems we should anyway.

mattip commented:
Waiting for a review ...

mattip commented:
ping, it would be nice to finish this up.

tylerjereddy commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1794,6 +1875,75 @@ make_full_arg_tuple(
+    }
+    return 0;
+
+fail:
```
There's only one `goto` directing here-- could remove it and just place above, but also clear enough this way and common in NumPy source I suppose. For a note of consistency, `_get_coredim_sizes` below just dumps the exception in the control flow without a goto.

mattip replied:
refactored

tylerjereddy commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2051,57 +2201,72 @@ _parse_axis_arg(PyUFuncObject *ufunc, int core_num_dims[], PyObject *axis,
+                int core_index = dim_offset + idim;
+                int core_dim_index = ufunc->core_dim_ixs[core_index];
+                npy_intp core_dim_size = core_dim_sizes[core_dim_index];
+                npy_intp op_dim_size;
```
Just to confirm -- does `op_dim_size` actually have to be declared inside this for loop instead of at the top of the function with other type declarations? So, limiting the scope of the variable type declaration within the for loop outweighs any cost associated with it happening iteratively because the compiler can optimize that out anyway?

mattip replied:
correct. It is a convenience. I find it easier to track variable lifetimes if they are defined in the block where they live, then I can forget about them after I grok the relevant code.

eric-wieser replied:
> outweighs any cost associated with

There is no runtime cost associated with moving a variable to an inner scope in C. C++ is a different story if your type has an expensive constructor / desctructor.

tylerjereddy commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -287,18 +287,85 @@ def test_all_ufunc(self):
         pass
 
     def test_signature(self):
+        # from include/numpy/ufuncobject.h
```
Maybe just a question of style, but I'd probably favor splitting the 78 line unit test function here into separately named test functions such that when they fail there's a more meaningful connection between the failure and the name of the test.

Parametrizing the tests from there may also further assist with meaningful test error tracebacks, but arguably this is just bonus stuff I suppose & these big "unit tests" are fairly common from what I've seen so far in NumPy source.

mattip replied:
Split them into separate tests

tylerjereddy commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -287,18 +287,85 @@ def test_all_ufunc(self):
+        assert_equal(ixs, (0, 1, 1, 2, 0, 2))
+        assert_equal(flags, (size_unset,)*3)
+        assert_equal(sizes, (-1, -1, -1))
+        # matmul signature from _umath_tests
```
some of the comments like this might be suitable locations for sudivision into new tests if you go that route

tylerjereddy commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -866,6 +927,89 @@ def test_innerwt_empty(self):
+        assert_raises(ValueError, umt.cross1d, a, np.arange(4.))
+        assert_raises(ValueError, umt.cross1d, a, np.arange(3.), np.zeros((3, 4)))
+
+    def test_can_ignore_signature(self):
```
Similar thought to above re: potential advantage of splitting into at least a few more specialized smaller tests and / or using some parametrization.

mattip replied:
This one is more difficult since each one does something slightly different and there is a natural progression from the simple to the complex, mangling the input differently at each step. Can we leave it for later?

eric-wieser replied:
I also think this would be clearer as separate tests. You'd end up with a handful more array initialization, but it would make it much clearer what is being tested.

mattip commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -696,15 +698,29 @@ PyUFunc_Type
+          PyUFunc_MaskedInnerLoopSelectionFunc *masked_inner_loop_selector;
           npy_uint32 *op_flags;
           npy_uint32 *iter_flags;
+          /* new in version 1 */
```
correct

eric-wieser replied:
>  aren't new in the code, but were missing in the docs?

Can you spin this kind of fix out into a separate PR?

mattip commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -764,6 +780,10 @@ PyUFunc_Type
 
+   .. c:member:: int PyUFuncObject.version
+
+       The version of this struct, currently set to 1
```
The intention is to increment the version number whenever the struct changes. What kind of test would you suggest? We do verify that the version number matches before accessing the new (for version 1) attributes.

Any change in the struct can be considered a major, backward-incompatible change so a minor version is not needed.

tylerjereddy replied:
Ok, sounds like something in the code is always going to break if you make a change to the `struct`--I was just going to suggest probing the number of structure members or something like that as an extra check on stability.

tylerjereddy commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -130,8 +130,8 @@ typedef struct _tagPyUFuncObject {
-        /* Used to be unused field 'check_return' */
-        int reserved1;
+        /* Used to be unused field 'check_return', repurposed in 1.16 */
+        int version;
```
Shouldn't this be a constant?  `const int version;`

tylerjereddy replied:
We use `const int` for pickle version types in NumPy source for example

mattip replied:
`const` has no meaning in `c`, it is a `c++` construct

tylerjereddy replied:
What? `const int x` is meaningful in C--in fact, it is not even the same type as `int x`. It has meaning to the compiler and if you try assigning to it later in the control flow it won't work.

Also, you can use `const` with pointers too, in a different sense, plenty of uses for `const` in C.

Here, I think it is pretty clear that the code shouldn't be re-assigning to the constant `version`, so it should be typed as a `const int`, not `int`

mattip replied:
correct. Fixing.

tylerjereddy replied:
Grepping in the code base shows this (and a few others):

`numpy/core/src/multiarray/methods.c:    const int version = 1;`

I guess we may have to be cautious because we're using a really old C-standard in the CI tests, and `const` didn't exist in the K&R / C89 days I think. But it is clearly in our source. Apparently you can also do `int const version = 1;` and some prefer that but our code base does the first one. Is that not working?

mattip replied:
Here the `version` is a field in a struct. While I am a bit ambivalent about the desire to make it `const`, I have added code to do this in 31a0d21d7.

mattip commented:
@mhvk needs a rebase and review fixes

mattip committed:
Merge 'master' into branch

mattip committed:
MAINT: cleanup from review

mattip committed:
MAINT: make PyUFuncObject.version a const int

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -339,6 +339,28 @@ _is_alnum_underscore(char ch)
+    npy_intp size = (npy_intp)strtoll(str, &stop, 10);
+    #if defined(_MSC_VER)
+        #undef strtoll
+    #endif
```
Just use `npy_strtoll`, which already encapsulates this

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -804,6 +824,51 @@ PyUFunc_Type
+
+   .. c:member:: int PyUFuncObject.core_num_dim_ix
+
+       Number of distinct core dimension names in the signature
```
A bigger issue of it being part of the public API is that `sizeof(PyUFuncObject)` changes, which is an ABI break for any code that tries to store `PyUFuncObject`s statically.

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -328,6 +408,7 @@ def test_signature(self):
         except ValueError:
             pass
 
+    def test_signature_failure3(self):
```
This test-splitting would be nice to extract to a smaller PR too

eric-wieser commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -336,12 +417,6 @@ def test_signature(self):
-        enabled, num_dims, ixs = umt.test_signature(2, 1, "(i1,i2),(J_1)->(_kAB)")
-        assert_equal(enabled, 1)
-        assert_equal(num_dims, (2, 1, 1))
-        assert_equal(ixs, (0, 1, 2, 3))
```
What happened to this test?

mattip replied:
it became `test_signature2`

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -479,6 +479,26 @@ _is_alnum_underscore(char ch)
+ */
+ static npy_intp
+ _get_size(const char* str)
+ {
```
Remote the leaing space on these three lines (and the last one)

mattip replied:
fixed

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -479,6 +479,26 @@ _is_alnum_underscore(char ch)
+#if defined(_MSC_VER)
+    npy_intp size = (npy_intp)_strtoi64(str, &stop, 10);
+#else
+    npy_intp size = (npy_intp)strtoll(str, &stop, 10);
```
`npy_strtoll`

mattip replied:
refactored things to make it available

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4644,6 +4851,39 @@ PyUFunc_FromFuncAndDataAndSignature(PyUFuncGenericFunction *func, void **data,
     PyUFuncObject *ufunc;
+    /*
+     * Use memcpy and an initialzed instance to set the newly allocated PyUFuncObject.
+       The version field is a constant and cannot be initialzed in C directly
```
`*((int *)&ufunc.version) = UFUNC_VERSION` would work, I think

mattip replied:
While it would work, it is [undefined behaviour](https://stackoverflow.com/questions/3709548/changing-const-value-in-c) according to the C standard. So why even make it `const` in the first place, if we ourselves do not respect the difference?

eric-wieser replied:
I doubt the `memcpy` is any better-defined.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -489,6 +509,9 @@ _get_end_of_name(const char* str, int offset)
     while (_is_alnum_underscore(str[ret])) {
         ret++;
     }
+    if (str[ret] == '?') {
```
Function comment should be updated to explain that `?` is included.

eric-wieser commented on doc/source/reference/c-api.generalized-ufuncs.rst:
```diff
@@ -127,34 +127,49 @@ The formal syntax of signatures is as follows::
++-------------+----------------------------+-----------------------------------+
+| vecmat      | ``(n),(n,p)->(p)``         | vector-matrix multiplication      |
++-------------+----------------------------+-----------------------------------+
+| matvec      | ``(m,n),(n)->(m)``         | matrix-vector multiplication      |
```
These three examples could be backported too, if we wanted

mattip replied:
Do you mean PR #11133?

eric-wieser replied:
No, I simply mean this doc change doesn't need to be part of this large PR

mattip replied:
ok, thanks

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1934,6 +2013,72 @@ make_full_arg_tuple(
 }
 
+/*
+ * Validate that operands have enough dimenions, accounting for
```
typo

mattip replied:
fixed

mattip committed:
MAINT: refactor to share npy_strto{u}ll

mattip committed:
MAINT: clarify internal function docstring

mattip committed:
MAINT: typo

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -479,8 +480,26 @@ _is_alnum_underscore(char ch)
+ * Convert a string into a number
+ */
+static npy_longlong
+_get_size(const char* str)
```
This function should return and intp, and verify that the result fits in an intp, given that it's always cast when called.

mattip replied:
done

mattip replied:
fixed

mhvk replied:
Seems like the signature is still `npy_int` rather than `npy_intp`?

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -479,8 +480,26 @@ _is_alnum_underscore(char ch)
 }
 
+NPY_NO_EXPORT npy_longlong
+npy_strtoll(const char *str, char **endptr, int base);
```
Would prefer that we move this to a header, rather than repeating the prototype

mattip replied:
leftover cruft removed

mhvk commented:
@mattip - thanks for taking over!  I do hope to look at this all as well, now that I'm back from my holidays, but with classes starting things may not be fast, so in the meantime do proceed...

mattip committed:
MAINT: rework wrapping and sharing common functions

eric-wieser commented on numpy/core/src/common/wrapper.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef _COMMON_WRAPPER_H
```
This file name is pretty cryptic - I'd hope there's a collection of similar functions already we can put this with. `numpyos.c` has `strtold`, which is pretty similar - perhaps it should go there.

mattip replied:
adopted

eric-wieser commented on numpy/core/src/umath/_umath_tests.c.src:
```diff
@@ -195,6 +197,41 @@ static void
+    INIT_OUTER_LOOP_3
+    npy_intp is1=steps[0], is2=steps[1], os = steps[2];
+    BEGIN_OUTER_LOOP_3
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
```
Better I think to introduce some locals
```C
@typ@ i1_x = *(@typ@ *)(args[0] + 0*is1);
@typ@ i1_y = *(@typ@ *)(args[0] + 1*is1);
@typ@ i1_z = *(@typ@ *)(args[0] + 2*is1);

@typ@ i2_x = *(@typ@ *)(args[1] + 0*is2);
@typ@ i2_y = *(@typ@ *)(args[1] + 1*is2);
@typ@ i2_z = *(@typ@ *)(args[1] + 2*is2);
```
Which avoids the mix of dereferencing and multiplication below.

mattip replied:
adopted

eric-wieser commented:
I've taken all the parts I remarked "would be nice as a separate CR", and made a CR for them at #11175

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -696,15 +698,29 @@ PyUFunc_Type
+          char *core_signature;
+          PyUFunc_TypeResolutionFunc *type_resolver;
+          PyUFunc_LegacyInnerLoopSelectionFunc *legacy_inner_loop_selector;
+          void *reserved2;
```
This member right here is our get-out clause to avoid ABI issues. Instead of making the struct any longer, we can use this to point to an arbitrary and private `ufunc_internal` object allocated on the heap. With that, we become free to add new members at any point

mhvk replied:
I'm not sure it is necessarily best to keep things private; I think ufunc writers would like to have access to whether dimensions are frozen, etc., in case the write, e.g., their own type resolver. (I've certainly made use of other numbers inferred from the signature in ufunc code.)
But as a variant on this: we could let this point to a new explicitly defined object that is version dependent (and gets a version-dependent name).

eric-wieser replied:
> I think ufunc writers would like to have access to [...]

Can we use the same approach as we do for ndarray, where we provide accessor functions?

mattip replied:
adopt a form of this idea, which achieves the goals of allowing memcpy of the old size while still using a single malloc

eric-wieser replied:
Hmm, thinking again I think there's no safe way to do this if someone uses a `memcpy`...

eric-wieser committed:
Merge branch 'master' into gufunc-signature-modification2

mattip committed:
merge master into gufunc-signature-modification2

mattip committed:
MAINT: move functions to numpyos.c, adjust accordingly

mattip commented:
merged #11950 into branch

mattip committed:
MAINT: refactor inner loop in test

mattip committed:
merge heads

eric-wieser commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -209,9 +228,21 @@ typedef struct _tagPyUFuncObject {
          */
         npy_uint32 iter_flags;
+
+        ufunc_extension s_extension;
```
This can't go here, as it changes the size of `PyUfunc_Object` which is an ABI change. I think we're forced to allocate it on the heap.

eric-wieser replied:
Hmm, actually that suggestion doesn't help

eric-wieser commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -111,6 +111,24 @@ typedef int (PyUFunc_MaskedInnerLoopSelectionFunc)(
+        /*
+         * for each distinct core dimension, a set of flags OR'd together
+         * e.g., UFUNC_CORE_DIM_CAN_IGNORE if signature has ?
+         *       UFUNC_CORE_DIM_SIZE_UNSET for non-frozen dimensions.
```
Why not just `UFUNC_CORE_DIM_SIZE_FROZEN` to mean the inverse?

mattip replied:
@mhvk any opinions? I could not find a place where we explicitly use the flag value except in test_signature

mhvk replied:
I had is as `FROZEN` initially, but changed it because that meant that `0` was something for which the code does not have to do anything - i.e., all flags mean that some special action has to be taken (in this case to get the dimension from the inputs). Of course, here it doesn't matter all that much size really the way this is signaled is by making the preset size negative. Still, personally I like the logic. (But I think it became a bit more logical in the follow-up PR, which allowed inner dimensions to be broadcast.)

mattip committed:
DOC: update for changes in d8a8e4536

eric-wieser commented:
I think there are some open questions about ABI breakage we need to answer here before we go ahead and change the size of any structs. #5888 may be relevant.

mattip commented:
A bit of archeology, to set the stage

Looking at issue #5888, and digging into history, it seems there is precedent for adding fields to public-facing structs. For instance the `hash` field was added to `PyArray_Descr` in PR #5343 (in April 2015) with very little fanfare and without updating `C_ABI_VERSION` (which becomes `NPY_ABI_VERSION`), but with the [comment](https://github.com/numpy/numpy/pull/5343#issuecomment-89685346):

> PyObjects should only be allocated dynamically and passed by reference (the exception is static definition of a PyTypeObject), so I don't know why a size change would be a problem. CPython can change the size of builtin objects from one version to another: does Cython complain about that?

The change to `PyArray_Descr` obliged us to depend on the cython warning suppression in PR #432 (in Sept 2012). The warning suppression was extended to `ndarray` in PR #5750 (April 2015), inadvertently removed recently in PR #10842 and restored in PR #11634.

mattip commented:
I do not think use of `memcpy` or other serialization of `PyUFuncObject` works in practice after `ufunc_dealloc`, even before this change. If someone does 
```
void * ufunc_small = malloc(sizeof(old_version_of_PyUFuncObject));
memcpy(ufunc_small, ufunc_big, sizeof(ufunc_small));
ufunc_dealloc(ufunc_big);
```
they have `free`ed a large number of pointers-to-arrays needed for correct functioning of the `PyUFuncObject`. 

If they avoid calling `ufunc_dealloc`, the `ufunc->extension` in d8a8e4536 will work after the `memcpy` just as `ufunc->core_num_dims` will work, it will access the array allocated for `ufunc_big`.

mattip committed:
DOC: warn about directly using PyUFuncObject and PyArray_Descr

eric-wieser commented:
> it seems there is precedent for adding fields to public-facing structs

I haven't had a chance to look through your references fully, but assuming that is the case - I don't think the `version` field is useful. Can you give me an example of when you'd inspect it?

mattip commented:
@eric-wieser at the beginning of `PyUfunc_GeneralizedFunction(ufunc, args, kwds, array_ops)` (which is used for ufuncs with generalized signatures) the `ufunc` data are matched to the `array_ops`' shapes to determine over which axes to iterate. Local variables are used to facilitate the matching. If somehow an older `ufunc` is passsed in, these variables are[ initialized to default values](https://github.com/mhvk/numpy/blob/gufunc-signature-modification2/numpy/core/src/umath/ufunc_object.c#L2501). If a current `ufunc` is passed in, the variables are initialized [using the values stored in the newer version of the ufunc](https://github.com/mhvk/numpy/blob/gufunc-signature-modification2/numpy/core/src/umath/ufunc_object.c#L2495).

mattip commented:
Now that we use the `extension` field (repurposed from `reserved2`) as well, perhaps the use of `version` is redundant, but it may be useful for some future change in the struct.

eric-wieser commented:
> If somehow an older ufunc is passsed in

Given the requirement that ufuncs are only ever allocated via one of our construction functions, how can this ever happen?

mhvk commented:
Wouldn't the `version` tell what the extension actually contained?

(Sorry, I really need to have a more thorough look at what this PR has become! Should have some time this week.)

eric-wieser commented:
I don't think that the `extension` provides any value now that we've decided that appending struct members is ok anyway.

eric-wieser commented:
I suppose the use case is someone compiling against new versions of the numpy headers and wanting access to the new struct members, but then linking against an old `multiarray`. In that case, I feel like the module should just ask for the multiarray module version, rather than spreading that version across all our objects.

mattip commented:
So what to do with this PR? Some options that jump to mind:

- **Merge as is.** We have no clear policy and no clear problem statement, so this PR is trying hard to protect itself from the unkown unkowns, in the hope that the extra work (a few strange fields and two c-level `if` checks) is worth the trouble.
- **Remove all version-related code.** We can safely assume no-one will ever create a `PyUFuncObject` in one runtime and use it in a different one, so just remove the extra checks and the hoops to use `version`, `extension` and `s_extension`, no version checking needed.
- **Set `version` to the `numpy/version.py` version**. When using the struct have some fast way of checking the version compatibility
- **Something else**

mhvk commented:
Sorry for still not coming back to this - I hope to soon. In the meantime, what was really wrong with the original idea, to just extend the struct and use a version number? If a pointer to an old struct is passed on, it will have version=0, and so the outer, uninitlialized parts will never be accessed. 

I do worry we're overthinking this - it should be possible to extend a structure without having to hide the extension...

mattip commented:
@eric-wieser - do you agree with @mhvk that we should go back to the situation before repurposing `reserved2` as `extension`? Should this hit the mailing list?

mattip commented:
I prefer 1 for now, maybe someday we can transition to an opaque API once cython/cython#2429 is closed (work is underway, see cython/cython#2640). Cython 0.29 and later will not emit the "size changed" warnings.

`PyUFuncObject` structs somehow created with an older runtime or headers will have the value 0. What do we put there instead in the newer runtime? `011000 =hex(1 << 24) | (16 << 16) | 0) like for [`PY_VERSION_HEX`](https://github.com/python/cpython/blob/master/Include/patchlevel.h#L28)?

eric-wieser commented:
> somehow created with an older runtime or headers will have the value 0

At risk of repeating myself above: Why do we need to worry about this? There can never be more than one numpy runtime per python runtime as the module names would be non-unique, and passing objects between python runtimes doesn't sound safe to me at all.

Given this, the only case that matters is end users compiled with the new headers who want to set / inspect the struct member - and they should decode if they can do so by using our "get runtime version" function, which I forget thr name of.

mattip commented:
> "get runtime version" function

There are 3 runtime versions: 

- the ABI version `NPY_ABI_VERSION` exposed as `PyArray_GetNDArrayCVersion` and `numpy._get_ndarray_c_version`, 
- the API version `NPY_API_VERSION` exposed as `PyArray_GetNDArrayCFeatureVersion`, and 
- the version from `./setup.py`'s MAJOR MINOR MICRO exposed various ways in `numpy.version`. 

Only the last one is relevant, since as far as I can tell changing the `PyObject` struct sizes has traditionally not changed the ABI nor API version. The motivation for PR #12074 is to make the third one accessable to C code.

mattip commented:
Yes, `NPY_API_VERSION` is incremented as you state. We crossposted, I asked for comments on the mailing list to PR #12074. What you are proposing is a valid alternative, and truthfully seems more appropriate.

mattip committed:
ENH: increment and use NPY_API_VERSION in PyUFuncObject->version

mattip commented:
Reworked to increment NPY_API_VERSION and use it instead of a private version number

eric-wieser commented on doc/release/1.16.0-notes.rst:
```diff
@@ -72,6 +72,8 @@ comparisons.
 =============
 
+The ``NPY_API_VERSION`` was incremented to 0x0000D since the `PyUFuncObject`'s
+size changed.
```
This is not the reason - it incremented because `PyUFuncObject.core_dim_flags` and `PyUFuncObject.core_dim_sizes` were added to the API.

mattip replied:
fixed

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -776,6 +796,10 @@ PyUFunc_Type
 
+   .. c:member:: int PyUFuncObject.version
+
+       The version of this struct, currently set to 1
```
This is outdated too now, right?

mattip replied:
fixed

mattip commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -713,10 +730,13 @@ PyUFunc_Type
           PyUFunc_TypeResolutionFunc *type_resolver;
           PyUFunc_LegacyInnerLoopSelectionFunc *legacy_inner_loop_selector;
-          void *reserved2;
+          ufunc_extension *extension;
```
fixed

mattip committed:
:DOC: update to reflect API changes

eric-wieser commented on numpy/core/include/numpy/ufuncobject.h:
```diff
@@ -130,8 +130,8 @@ typedef struct _tagPyUFuncObject {
-        /* Used to be unused field 'check_return' */
-        int reserved1;
+        /* Used to be unused field 'check_return', repurposed in 1.16 */
+        const int version;
```
This is outdated now, right?

mattip replied:
In `PyUFunc_FromFuncAndDataAndSignature` it is set to `NPY_API_VERSION`, which is then checked in `PyUFunc_GeneralizedFunction` (via a call to `_initialize_variable_parts`) to make sure it is `>= 0x00000d`

eric-wieser replied:
This seems pointless to me

mattip replied:
Removing. If it becomes an issue we can revisit this.

eric-wieser commented on numpy/core/src/umath/_umath_tests.c.src:
```diff
@@ -285,6 +326,26 @@ static void
+    Ufunc(2, 1, None_,
+        r'''inner on the last dimension and broadcast on the rest \n"
+        "     \"(i),(i)->()\" \n''',
+          TD('ld'),
```
something weird going on with the indentation here

mattip replied:
fixed, and aligned next block on 4-space-indent as well

eric-wieser commented on numpy/core/src/umath/_umath_tests.c.src:
```diff
@@ -295,7 +356,9 @@ static char innerwt_signatures[] = { NPY_LONG, NPY_LONG, NPY_LONG, NPY_LONG, NPY
 static char matrix_multiply_signatures[] = { NPY_LONG, NPY_LONG, NPY_LONG,  NPY_FLOAT, NPY_FLOAT, NPY_FLOAT,  NPY_DOUBLE, NPY_DOUBLE, NPY_DOUBLE };
-
+static PyUFuncGenericFunction cross1d_functions[] = { LONG_cross1d, DOUBLE_cross1d };
+static void * cross1d_data[] = { (void *)NULL, (void *)NULL };
```
nit: No space after `*` to match the declaration above

mattip replied:
reworked pointer declarations `* name` -> `*name`

eric-wieser commented:
Thanks for iterating on the versioning problem.

I've one remaining comment, which I think I mentioned before, but can't find. I think we should invert `UFUNC_CORE_DIM_SIZE_UNSET` to `UFUNC_CORE_DIM_FIXED_SIZE` (or `FROZEN_SIZE`), since that means:
* setting the flags to `0` is a sensible default
* we don't end up with double negatives

mattip committed:
MAINT: formatting, remove version, rework flags

mattip commented:
reformatted, reworked flag, removed version from struct and docs.

mhvk commented:
Piping in from what has become the sidelines: I actually chose the flag with some care, in that if a flag is set, it implies the code needs to do extra work. And a constant size is the obviously simpler case ;-) If the double negative is a problem, it could be `UFUNC_CORE_DIM_SIZE_FREE`.

Though I will add that the logic become truly obvious only when I added the broadcasting, when I also had `UFUNC_CORE_DIM_CAN_BROADCAST`.

mattip commented:
I have no particular preference. Note the flag is never read in the code itself, only in tests. Perhaps removing it is another way to resolve the discussion, until the broadcast proposal arises again.

mhvk commented:
@mattip - I think I'd prefer to keep the flag since there is the other one for `matmul` as well. I also prefer to keep the sense that flag set means work - how about `UFUNC_CORE_DIM_FLEXIBLE_SIZE` or `...SIZE_FREE` - it suggests that the underlying code can handle a non-fixed size.

eric-wieser commented:
> I also prefer to keep the sense that flag set means work

That's perhaps a reasonable interpretation.

I don't think it matters too much - the only reason I mention it is this is becoming our public API, so once we choose we're stuck with it.

eric-wieser commented:
How about `UFUNC_CORE_DIM_SIZE_INFERRED`, which does not form a double negative, and indicates work in the way @mhvk  mentions?

mattip commented:
Updated to UFUNC_CORE_DIM_SIZE_INFERRED, and enhanced documentation to cross-reference `frozen`, learning about [arbitrary anchors in rst](http://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#ref-role) in the process

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -870,6 +894,21 @@ PyUFunc_Type
+   .. c:member:: npy_intp *PyUFuncObject.core_dim_sizes
+
+       For each distinct core dimension, the possible
+       :ref:`frozen <frozen>` size (``-1`` if not frozen)
```
Why are we using both `-1` and `UFUNC_CORE_DIM_SIZE_INFERRED` to indicate this?

mattip replied:
@mhvk ?

mhvk replied:
I think that's a leftover from the initial work by Jaime. I think it is fine to remove the guarantee that it is `-1` if not frozen, as indeed the flag already indicates that (and perhaps we want to use negative numbers to indicate something else in the future).

eric-wieser commented on numpy/core/code_generators/cversions.txt:
```diff
@@ -43,3 +43,5 @@
 # PyArray_SetWritebackIfCopyBase and deprecated PyArray_SetUpdateIfCopyBase.
 0x0000000c = a1bc756c5782853ec2e3616cf66869d8
 
+# Version 13 (Numpy 1.16) Size of PyUFuncObject changed
```
Again, this should mention the new struct member names

mattip committed:
DOC: tweak docs from review

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -685,7 +695,14 @@ PyUFunc_Type
+
+   - Never declare a non-pointer instance of the struct
+   - Never perform pointer arithmatic
+   - Never use ``sizof(PyUFuncObject)``
```
typo

mattip replied:
fixed

eric-wieser commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -685,7 +695,14 @@ PyUFunc_Type
+   of NumPy. To ensure compatibility:
+
+   - Never declare a non-pointer instance of the struct
+   - Never perform pointer arithmatic
```
typo: arithmetic

mattip replied:
fixed

mattip commented:
Fixed documentation review issues.

mhvk commented on doc/source/reference/c-api.types-and-structures.rst:
```diff
@@ -698,7 +715,7 @@ PyUFunc_Type
           void **data;
           int ntypes;
-          int reserved1;
+          int version;
```
I thought we were leaving this as `reserved1` (at least that is what is described below).

mattip replied:
reverted

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2429,72 +2609,41 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     }
     /*
-     * If keepdims is set and true, signal all dimensions will be the same.
+     * If keepdims is set and true, which means all input dimensions are
```
Either "which" -> "this", or better, replace the period with a comma (and lower-case "signal")

mattip replied:
changed

mhvk reviewed (commented):
Only trivia.

mattip committed:
MAINT: changes from review

mattip commented:
fixed comments from @mhvk

mattip commented:
@mhvk, @eric-wieser anything I can do to move this forward?

mhvk commented:
@mattip - you fixed my trivial comments, so I'm happy; thanks for carrying this on, after me hijacking it.

mattip commented:
Thanks @mhvk, @eric-wieser for the patience. I merged this even though by now I am probably more considered a contributor than a reviewer, after consulting with core developers in the weekly status meeting.

mhvk commented:
@mattip - thank you for first starting, then commenting, and finally shepparding this on!  And thanks, @eric-wieser, for the as always very useful comments/critique!

I do hope to get back to working on speeding up the ufuncs... Though perhaps playing with `__array_function__` first.
